.github/workflows/main.yml

```diff
@@ -0,0 +1,135 @@
+      CARGO_INCREMENTAL: 0
+    strategy:
+      matrix:
+        include:
+        - os: ubuntu-latest
+          rust: stable
+          other: i686-unknown-linux-gnu
+        - os: ubuntu-latest
+          rust: beta
+          other: i686-unknown-linux-gnu
+        - os: ubuntu-latest
+          rust: nightly
+          other: i686-unknown-linux-gnu
+        - os: macos-latest
+          rust: stable
+          other: x86_64-apple-ios
+        - os: windows-latest
+          rust: stable-msvc
+          other: i686-pc-windows-msvc
+        - os: windows-latest
+          rust: beta-gnu
+          other: i686-pc-windows-gnu
+    steps:
+    - uses: actions/checkout@v2
+    - run: rustup update --no-self-update ${{ matrix.rust }} && rustup default ${{ matrix.rust }}
+    - run: rustup target add ${{ matrix.other }}
+    - run: rustup component add rustc-dev llvm-tools-preview rust-docs
+      if: matrix.rust == 'nightly'
+    - run: sudo apt update -y && sudo apt install gcc-multilib -y
+      if: matrix.os == 'ubuntu-latest'
+    - run: rustup component add rustfmt || echo "rustfmt not available"
+
+    # Deny warnings on CI to avoid warnings getting into the codebase, and note
+    # the `force-system-lib-on-osx` which is intended to fix compile issues on
+    # OSX where compiling curl from source on OSX yields linker errors on Azure.
+    #
+    # Note that the curl issue is traced back to alexcrichton/curl-rust#279
+    # where it looks like the OSX version we're actually running on is such that
+    # a symbol is emitted that's never worked. For now force the system library
+    # to be used to fix the link errors.
+    - run: cargo test --features 'deny-warnings curl/force-system-lib-on-osx'
+    - run: cargo test -p cargo-test-support
+    - run: cargo test -p cargo-platform
+
+  resolver:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - run: rustup update stable && rustup default stable
+    - run: cargo test --manifest-path crates/resolver-tests/Cargo.toml
+
+  build_std:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - run: rustup update nightly && rustup default nightly
+    - run: cargo test --manifest-path crates/resolver-tests/Cargo.toml
+    - run: rustup component add rust-src
+    - run: cargo build
+    - run: cargo test --test build-std
+      env:
+        CARGO_RUN_BUILD_STD_TESTS: 1
+  docs:
+    env:
+      MDBOOK_VERSION: 0.3.7
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - run: rustup update nightly && rustup default nightly
+    - run: rustup component add rust-docs
+
+    # Install mdbook, cached
+    - uses: actions/cache@v1
+      with:
+        path: ${{ runner.tool_cache }}/mdbook
+        key: mdbook-${{ env.MDBOOK_VERSION }}
+    - run: echo "::add-path::${{ runner.tool_cache }}/mdbook/bin"
+    - run: cargo install --root ${{ runner.tool_cache }}/mdbook --version ${{ env.MDBOOK_VERSION }} mdbook
```

Comment:
Is there a reason to do it this way instead of downloading the pre-built executable?  Downloading should generally be much faster.

Reply:
I thought we'd had issues in the past where sometimes mdbook didn't have a precompiled binary available, but I may be misremembering. I'd also already seen all the code necessary to cache `cargo install` on GitHub Actions with Wasmtime where one builder runs `cargo audit`, so I figured it wouldn't be too bad to update that here. This also in theory makes it somewhat easy to update the version to.

I don't mind preferring either way though.

Reply:
I would prefer downloading the pre-built binary directly. I've been managing the releases for the past year, so they should be semi-reliable for now. 

I semi-remember that GHA cache works weird with PRs, and there is no way to clear the cache if there is an issue.